File: src/main/java/uk/gov/justice/hmpps/prison/aop/AuthorisationAspect.java
package uk.gov.justice.hmpps.prison.aop;

import lombok.extern.slf4j.Slf4j;
import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.aspectj.lang.annotation.Pointcut;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.stereotype.Component;
import uk.gov.justice.hmpps.prison.security.AuthenticationFacade;
import uk.gov.justice.hmpps.prison.security.VerifyAgencyAccess;
import uk.gov.justice.hmpps.prison.security.VerifyBookingAccess;
import uk.gov.justice.hmpps.prison.security.VerifyOffenderAccess;
import uk.gov.justice.hmpps.prison.service.AgencyService;
import uk.gov.justice.hmpps.prison.service.BookingService;
import uk.gov.justice.hmpps.prison.service.support.AgencyRequest;

@Aspect
@Slf4j
@Component
public class AuthorisationAspect {
    private final BookingService bookingService;
    private final AgencyService agencyService;

    public AuthorisationAspect(final BookingService bookingService, final AgencyService agencyService) {
        this.bookingService = bookingService;
        this.agencyService = agencyService;
    }

    @Pointcut("@annotation(uk.gov.justice.hmpps.prison.security.VerifyOffenderAccess) && execution(* *(String,..)) && args(offenderNo,..)")
    public void verifyOffenderAccessPointcut(final String offenderNo) {
        // no code needed - pointcut definition
    }

    @Pointcut("@annotation(uk.gov.justice.hmpps.prison.security.VerifyBookingAccess) && execution(* *(Long,..)) && args(bookingId,..)")
    public void verifyBookingAccessPointcut(final Long bookingId) {
        // no code needed - pointcut definition
    }

    @Pointcut("@annotation(uk.gov.justice.hmpps.prison.security.VerifyAgencyAccess) && execution(* *(String,..)) && args(agencyId,..)")
    public void verifyAgencyAccessPointcut(final String agencyId) {
        // no code needed - pointcut definition
    }

    @Pointcut("@annotation(uk.gov.justice.hmpps.prison.security.VerifyAgencyAccess) && args(uk.gov.justice.hmpps.prison.service.support.AgencyRequest,..) && args(request,..)")
    public void verifyAgencyRequestAccessPointcut(final AgencyRequest request) {
        // no code needed - pointcut definition
    }

    @Before(value = "verifyBookingAccessPointcut(bookingId)", argNames = "jp,bookingId")
    public void verifyBookingAccess(final JoinPoint jp, final Long bookingId) {
        log.debug("Verifying booking access for booking [{}]", bookingId);

        final var signature = (MethodSignature) jp.getSignature();
        final var method = signature.getMethod();
        final var annotation = method.getAnnotation(VerifyBookingAccess.class);
        final var overrideRoles = annotation.overrideRoles();

        if (AuthenticationFacade.hasRoles(overrideRoles)) {
            bookingService.checkBookingExists(bookingId);
        } else {
            bookingService.verifyBookingAccess(bookingId, overrideRoles);
        }
    }

    @Before(value = "verifyOffenderAccessPointcut(offenderNo)", argNames = "jp,offenderNo")
    public void verifyOffenderAccess(final JoinPoint jp, final String offenderNo) {
        log.debug("Verifying offender access for offender No [{}]", offenderNo);
        final var signature = (MethodSignature) jp.getSignature();
        final var method = signature.getMethod();
        final var annotation = method.getAnnotation(VerifyOffenderAccess.class);
        final var overrideRoles = annotation.overrideRoles();

        bookingService.verifyCanViewSensitiveBookingInfo(offenderNo, overrideRoles);
    }

    @Before(value = "verifyAgencyAccessPointcut(agencyId)", argNames = "jp,agencyId")
    public void verifyAgencyAccess(final JoinPoint jp, final String agencyId) {
        log.debug("Verifying agency access for agency [{}]", agencyId);

        if (AuthenticationFacade.hasRoles(getOverrideRoles(jp))) {
            agencyService.checkAgencyExists(agencyId);
        } else {
            agencyService.verifyAgencyAccess(agencyId);
        }
    }

    @Before(value = "verifyAgencyRequestAccessPointcut(request)", argNames = "jp,request")
    public void verifyAgencyRequestAccess(final JoinPoint jp, final AgencyRequest request) {
        log.debug("Verifying agency access for agency [{}]", request.getAgencyId());

        if (AuthenticationFacade.hasRoles(getOverrideRoles(jp))) {
            agencyService.checkAgencyExists(request.getAgencyId());
        } else {
            agencyService.verifyAgencyAccess(request.getAgencyId());
        }
    }

    private String[] getOverrideRoles(final JoinPoint jp) {
        final var signature = (MethodSignature) jp.getSignature();
        final var method = signature.getMethod();
        final var annotation = method.getAnnotation(VerifyAgencyAccess.class);
        return annotation.overrideRoles();
    }

}

File: src/main/java/uk/gov/justice/hmpps/prison/aop/AuthorisationAspect.kt
package uk.gov.justice.hmpps.prison.aop

import org.aspectj.lang.JoinPoint
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Before
import org.aspectj.lang.annotation.Pointcut
import org.aspectj.lang.reflect.MethodSignature
import org.slf4j.LoggerFactory
import org.springframework.security.access.AccessDeniedException
import org.springframework.stereotype.Component
import uk.gov.justice.hmpps.prison.repository.jpa.model.StaffUserAccount
import uk.gov.justice.hmpps.prison.repository.jpa.repository.StaffUserAccountRepository
import uk.gov.justice.hmpps.prison.security.AuthenticationFacade
import uk.gov.justice.hmpps.prison.security.VerifyAgencyAccess
import uk.gov.justice.hmpps.prison.security.VerifyBookingAccess
import uk.gov.justice.hmpps.prison.security.VerifyOffenderAccess
import uk.gov.justice.hmpps.prison.security.VerifyStaffAccess
import uk.gov.justice.hmpps.prison.service.AgencyService
import uk.gov.justice.hmpps.prison.service.BookingService
import uk.gov.justice.hmpps.prison.service.support.AgencyRequest

@Aspect
@Component
class AuthorisationAspect(
  private val bookingService: BookingService,
  private val agencyService: AgencyService,
  private val authenticationFacade: AuthenticationFacade,
  private val staffUserAccountRepository: StaffUserAccountRepository,
) {
  @Pointcut("@annotation(uk.gov.justice.hmpps.prison.security.VerifyOffenderAccess) && execution(* *(String,..)) && args(offenderNo,..)")
  fun verifyOffenderAccessPointcut(offenderNo: String) {
    // no code needed - pointcut definition
  }

  @Pointcut("@annotation(uk.gov.justice.hmpps.prison.security.VerifyBookingAccess) && execution(* *(Long,..)) && args(bookingId,..)")
  fun verifyBookingAccessPointcut(bookingId: Long) {
    // no code needed - pointcut definition
  }

  @Pointcut("@annotation(uk.gov.justice.hmpps.prison.security.VerifyAgencyAccess) && execution(* *(String,..)) && args(agencyId,..)")
  fun verifyAgencyAccessPointcut(agencyId: String) {
    // no code needed - pointcut definition
  }

  @Pointcut("@annotation(uk.gov.justice.hmpps.prison.security.VerifyAgencyAccess) && args(uk.gov.justice.hmpps.prison.service.support.AgencyRequest,..) && args(request,..)")
  fun verifyAgencyRequestAccessPointcut(request: AgencyRequest) {
    // no code needed - pointcut definition
  }

  @Pointcut("@annotation(uk.gov.justice.hmpps.prison.security.VerifyStaffAccess) && execution(* *(Long,..)) && args(staffId,..)")
  fun verifyStaffAccessPointcut(staffId: Long) {
    // no code needed - pointcut definition
  }

  @Before(value = "verifyBookingAccessPointcut(bookingId)", argNames = "jp,bookingId")
  fun verifyBookingAccess(jp: JoinPoint, bookingId: Long) {
    log.debug("Verifying booking access for booking [{}]", bookingId)
    val signature = jp.signature as MethodSignature
    val method = signature.method
    val annotation = method.getAnnotation(VerifyBookingAccess::class.java)
    val overrideRoles = annotation.overrideRoles
    val accessDeniedError = annotation.accessDeniedError
    if (AuthenticationFacade.hasRoles(*overrideRoles)) {
      bookingService.checkBookingExists(bookingId)
    } else {
      bookingService.verifyBookingAccess(bookingId, accessDeniedError, *overrideRoles)
    }
  }

  @Before(value = "verifyOffenderAccessPointcut(offenderNo)", argNames = "jp,offenderNo")
  fun verifyOffenderAccess(jp: JoinPoint, offenderNo: String) {
    log.debug("Verifying offender access for offender No [{}]", offenderNo)
    val signature = jp.signature as MethodSignature
    val method = signature.method
    val annotation = method.getAnnotation(VerifyOffenderAccess::class.java)
    val overrideRoles = annotation.overrideRoles
    val accessDeniedError = annotation.accessDeniedError
    bookingService.getOffenderIdentifiers(offenderNo, accessDeniedError, *overrideRoles)
  }

  @Before(value = "verifyAgencyAccessPointcut(agencyId)", argNames = "jp,agencyId")
  fun verifyAgencyAccess(jp: JoinPoint, agencyId: String) {
    log.debug("Verifying agency access for agency [{}]", agencyId)
    val signature = jp.signature as MethodSignature
    val method = signature.method
    val annotation = method.getAnnotation(VerifyAgencyAccess::class.java)
    val overrideRoles = annotation.overrideRoles
    val accessDeniedError = annotation.accessDeniedError
    if (AuthenticationFacade.hasRoles(*overrideRoles)) {
      agencyService.checkAgencyExists(agencyId)
    } else {
      agencyService.verifyAgencyAccess(agencyId, accessDeniedError)
    }
  }

  @Before(value = "verifyAgencyRequestAccessPointcut(request)", argNames = "jp,request")
  fun verifyAgencyRequestAccess(jp: JoinPoint, request: AgencyRequest) {
    verifyAgencyAccess(jp, request.agencyId)
  }

  @Before(value = "verifyStaffAccessPointcut(staffId)", argNames = "jp,staffId")
  fun verifyStaffAccess(jp: JoinPoint, staffId: Long) {
    log.debug("Verifying staffId access for staffId [{}]", staffId)
    val signature = jp.signature as MethodSignature
    val method = signature.method
    val annotation = method.getAnnotation(VerifyStaffAccess::class.java)
    val overrideRoles = annotation.overrideRoles
    if (!AuthenticationFacade.hasRoles(*overrideRoles)) {
      val currentUsername: String = authenticationFacade.getCurrentUsername()
        ?: throw AccessDeniedException("No current username for staffId=$staffId")
      staffUserAccountRepository.findByUsername(currentUsername)
        .ifPresentOrElse(
          { staffUserAccount: StaffUserAccount ->
            val userStaffId = staffUserAccount.staff.staffId
            if (userStaffId != staffId) {
              throw AccessDeniedException("staff=$userStaffId accessing details of other staff=$staffId")
            }
          },
          {
            throw AccessDeniedException("Cannot find staff id for username=$currentUsername")
          },
        )
    }
  }

  private companion object {
    private val log = LoggerFactory.getLogger(this::class.java)
  }
}
